File: c#/builder/builder/XmlFormBuilder.cs
using System;

namespace builder
{
	public class XmlFormBuilder : FormBuilder
	{
		public Object StartBuilding() {
			Console.WriteLine("<form>");
			return 1;
		}

		public void EndBuilding() {
			Console.WriteLine("</form>");
		}

		public Object StartGroup(Object parent, string label) {
			int level = Indent(parent);
			Console.WriteLine("<group label=\"{0}\">", label);
			return level + 1;
		}

		public void EndGroup(Object parent) {
			Indent(parent);
			Console.WriteLine("</group>");
		}

		public void BuildCheckbox(Object parent, string label) {
			Indent(parent);
			Console.WriteLine("<checkbox label=\"{0}\" />", label);
		}

		public Object StartRadioGroup(Object parent) {
			int level = Indent(parent);
			Console.WriteLine("<radio>");
			return level + 1;
		}

		public void BuildRadioButton(Object radioGroup, string label) {
			Indent(radioGroup);
			Console.WriteLine("<button label=\"{0}\" />", label);
		}

		public void EndRadioGroup(Object parent) {
			Indent(parent);
			Console.WriteLine("</radio>");
		}

		public void BuildText(Object parent, string label) {
			Indent(parent);
			Console.WriteLine("<text label=\"{0}\" />", label);
		}

		private int Indent(Object levelObj) {
			int level = (int) levelObj;
			for (int i = 0; i < level; i++) {
				Console.Write("\t");
			}
			return level;
		}
	}
}

File: various/oom-lab1/XmlFormBuilder.java
package lab1;

public class XmlFormBuilder implements FormBuilder {

	public Object startBuilding() {
		System.out.println("<form>");
		return 1;
	}

	public void endBuilding() {
		System.out.println("</form>");
	}

	public Object startGroup(Object parent, String label) {
		int indentationLevel = (Integer) parent;
		indent(indentationLevel);
		System.out.println("<group label=\"" + label + "\" />");
		return indentationLevel + 1;
	}

	public void endGroup(Object parent) {
		int indentationLevel = (Integer) parent;
		indent(indentationLevel);
		System.out.println("</group>");
	}

	public Object startRadioGroup(Object parent) {
		Integer indentationLevel = (Integer) parent;
		indent(indentationLevel);
		System.out.println("<radio>");
		return indentationLevel + 1;
	}

	public void endRadioGroup(Object parent) {
		int indentationLevel = (Integer) parent;
		indent(indentationLevel);
		System.out.println("</radio>");
	}

	public void buildCheckbox(Object parent, String label) {
		Integer indentationLevel = (Integer) parent;
		indent(indentationLevel);
		System.out.println("<checkbox label=\"" + label + "\" />");
	}

	public void buildRadioButton(Object parent, String label) {
		Integer indentationLevel = (Integer) parent;
		indent(indentationLevel);
		System.out.println("<radio label=\"" + label + "\" />");
	}

	public void buildText(Object parent, String label) {
		Integer indentationLevel = (Integer) parent;
		indent(indentationLevel);
		System.out.println("<text label=\"" + label + "\" />");
	}

	private void indent(int level) {
		StringBuilder indentation = new StringBuilder();
		for (int i = 0; i < level; i++) {
			indentation.append("\t");
		}
		System.out.print(indentation);
	}
}
